File: packages/api/amplify_api_ios/ios/Classes/Auth/FlutterAuthProvider.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import Foundation
import Amplify
import AmplifyPlugins
import AWSPluginsCore
import Flutter

extension FlutterError: Error {}

/// A factory of [FlutterAuthProvider] instances. Manages shared state for all providers.
class FlutterAuthProviders: APIAuthProviderFactory {
    /// Thread to perform wait activities on.
    private static let queue = DispatchQueue(label: "FlutterAuthProviders")

    /// Retrieves the latest token for `type` by calling into Flutter via the plugin's method channel.
    static func getToken(for type: AWSAuthorizationType) -> Result<String, Error> {
        
        let unknownError: Result<String, Error> = .failure(
            APIError.unknown("Token could not be retrieved",
                             "An unknown error occurred.",
                             nil))

        // Preventative measure so as to not block the main thread. There is no expectation this
        // will be true but it was seen on Android during REST configuration.
        if Thread.isMainThread {
            return unknownError
        }

        var token: Result<String, Error>?
        queue.sync {
            let completer = DispatchSemaphore(value: 0)

            DispatchQueue.main.async {
                SwiftAmplifyApiPlugin.nativeApiPlugin.getLatestAuthTokenProviderName(type.rawValue) { resultToken, error in
                    defer { completer.signal() }
                    
                    if let resultToken = resultToken {
                        token = .success(resultToken)
                    } else {
                        let error = error as? FlutterError
                        token = .failure(APIError.operationError(
                            "Unable to retrieve token for \(type)",
                            """
                            Make sure you register your auth providers in the addPlugin call and \
                            that getLatestAuthToken returns a value.
                            """,
                            error))
                    }
                }
            }

            let timeout: DispatchTimeInterval = .seconds(5)
            let waitResult = completer.wait(timeout: .now() + timeout)
            if waitResult == .timedOut {
                token = .failure(APIError.operationError(
                    "Token retrieval timed out after \(timeout)",
                    """
                    Please retry the call and make sure your getLatestAuthToken function completes.
                    """))
            }
        }

        return token ?? unknownError
    }

    private let authProviders: Set<AWSAuthorizationType>

    init(_ authProviders: [AWSAuthorizationType]) {
        self.authProviders = Set(authProviders)
    }

    override func oidcAuthProvider() -> AmplifyOIDCAuthProvider? {
        guard authProviders.contains(.openIDConnect) else {
            return nil
        }
        return FlutterAuthProvider(type: .openIDConnect)
    }

    override func functionAuthProvider() -> AmplifyFunctionAuthProvider? {
        guard authProviders.contains(.function) else {
            return nil
        }
        return FlutterAuthProvider(type: .function)
    }
}

/// A provider which manages token retrieval for its [AWSAuthorizationType].
struct FlutterAuthProvider: AmplifyOIDCAuthProvider, AmplifyFunctionAuthProvider {
    let type: AWSAuthorizationType

    func getLatestAuthToken() -> Result<String, Error> {
        return FlutterAuthProviders.getToken(for: type)
    }
}

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/FlutterAuthProvider.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazonaws.amplify.amplify_datastore

import android.os.Looper
import com.amazonaws.amplify.amplify_datastore.exception.ExceptionMessages
import com.amazonaws.amplify.amplify_datastore.pigeons.NativeApiPlugin
import com.amplifyframework.api.ApiException
import com.amplifyframework.api.aws.ApiAuthProviders
import com.amplifyframework.api.aws.AuthorizationType
import com.amplifyframework.api.aws.sigv4.FunctionAuthProvider
import com.amplifyframework.api.aws.sigv4.OidcAuthProvider
import io.flutter.Log
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout

/**
 * Manages the shared state of all [FlutterAuthProvider] instances.
 */
class FlutterAuthProviders(
    private val authProviders: List<AuthorizationType>,
    private val nativeApiPlugin: NativeApiPlugin
) {
    private companion object {
        /**
         * Timeout on a single [getToken] call.
         */
        const val getTokenTimeoutMillis = 5000L

        /**
         * Logger tag.
         */
        const val tag = "FlutterAuthProviders"

        /**
         * Name for suspending block in [getToken]. Used for debugging
         */
        val coroutineName = CoroutineName(tag)
    }

    /**
     * A factory of [FlutterAuthProvider] instances.
     */
    val factory: ApiAuthProviders by lazy {
        val authProviders = this.authProviders.toSet()
        val builder = ApiAuthProviders.Builder()
        if (authProviders.contains(AuthorizationType.AWS_LAMBDA)) {
            builder.functionAuthProvider(FlutterAuthProvider(this, AuthorizationType.AWS_LAMBDA))
        }
        if (authProviders.contains(AuthorizationType.OPENID_CONNECT)) {
            builder.oidcAuthProvider(FlutterAuthProvider(this, AuthorizationType.OPENID_CONNECT))
        }
        builder.build()
    }

    /**
     * Retrieves the token for [authType] or `null`, if unavailable.
     *
     * This function is typically called from within the Amplify library and from a thread besides
     * the main one, where it is safe to block.
     */
    fun getToken(authType: AuthorizationType): String? {
        // Not blocking the main thread is required for making platform channel calls without
        // deadlock.
        if (Thread.currentThread() == Looper.getMainLooper().thread) {
            Log.e(tag, ExceptionMessages.createGithubIssueString)
            return null
        }

        try {
            return runBlocking(coroutineName) {
                val completer = Job()

                val result = object : MethodChannel.Result {
                    var token: String? = null

                    override fun success(result: Any?) {
                        token = result as? String
                        launch(Dispatchers.Main) {
                            completer.complete()
                        }
                    }

                    override fun error(
                        errorCode: String,
                        errorMessage: String?,
                        errorDetails: Any?
                    ) {
                        launch(Dispatchers.Main) {
                            completer.complete()
                        }
                    }

                    override fun notImplemented() {
                        launch(Dispatchers.Main) {
                            completer.complete()
                        }
                    }
                }
                launch(Dispatchers.Main) {

                    nativeApiPlugin.getLatestAuthToken(authType.name) { resultToken ->
                        result.success(resultToken)
                    }
                }

                withTimeout(getTokenTimeoutMillis) {
                    completer.join()
                }

                return@runBlocking result.token
            }
        } catch (e: Exception) {
            Log.e(tag, "Exception in getToken", e)
            return null
        }
    }
}

/**
 * A provider which manages token retrieval for its [AuthorizationType].
 */
class FlutterAuthProvider(
    private val provider: FlutterAuthProviders,
    private val type: AuthorizationType
) : FunctionAuthProvider,
    OidcAuthProvider {
    private companion object {
        /**
         * Thrown when there is no token available for [type].
         */
        fun noTokenAvailable(type: AuthorizationType) = ApiException.ApiAuthException(
            "Unable to retrieve token for $type",
            """
                Make sure you register your auth providers in the addPlugin call and
                that getLatestAuthToken returns a value.
            """.trimIndent()
        )
    }

    override fun getLatestAuthToken(): String =
        provider.getToken(type) ?: throw noTokenAvailable(type)
}
